### PR TITLE
builder: Include EC2 Capacity Reservations in the aws builder

### LINKED ISSUE #255: Add support for EC2 Capacity Reservations
#### Community Note

Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request.
If you are interested in working on this issue or have submitted a pull request, please leave a comment.

#### Description

AWS launched on-demand capacity reservations and it seems like the amazon plugin for packer does not support specifying the reservation when creating an AMI via packer.

#### Use Case(s)

On-demand capacity for GPU based instances is very difficult to obtain, so we use capacity reservations which sets an instance of a specific type aside for us to use.

#### Potential References

https://aws.amazon.com/about-aws/whats-new/2021/10/amazon-ec2-capacity-reservation-fleet-capacity-reservations/

### PR DESCRIPTION
**Pull request by mihok**
### Changes

Adding support for the `CapacityReservationSpecification` in the `RunInstancesInput ` struct so that Packer can use CapacityReservations when building AMIs. Also some small formatting changes

Closes #255

### CODE REVIEW

**Comment by mihok** [comment]
I'm trying to figure out how to run the packer-sdc, it seems to panic when I try to generate anything, and `go generate builder/ebs/` doesnt seem to to actually generate anything. Help here is very appreciated!

This is the panic I receive, I'm not familiar enough with the project to know if this is something I did or am doing wrong:
```
> packer-sdc mapstructure-to-hcl2 -output . -type Config
panic: runtime error: slice bounds out of range [:-1]

goroutine 1 [running]:
github.com/hashicorp/packer-plugin-sdk/cmd/packer-sdc/internal/mapstructure-to-hcl2.(*Command).Run(0xc000120500, {0xc0000241a0, 0x4, 0x4})
	/home/mihok/development/pkg/mod/github.com/hashicorp/packer-plugin-sdk@v0.3.1/cmd/packer-sdc/internal/mapstructure-to-hcl2/mapstructure-to-hcl2.go:99 +0x1afa
```

**Comment by nywilken** [comment]
Hi @mihok I see that all the generated partials have been updated with the newly added fields. So I'm guessing you've managed to get pass the packer-sdc command. Is that correct? 

For the plugins we don't run the command directly. We run `go generate ./...` in the project root directory to let the tooling execute the packer-sdc command on our behalf. There is some pathing related logic that assumes it is being run through go generate. I hope that helps.

**Review by nywilken — CHANGES_REQUESTED** [review]

File: docs-partials/builder/ebs/Config-not-required.mdx [hunk 1]
```diff
@@ -9,6 +9,15 @@
   here may vary depending on the type of VM you use. See the
   [BlockDevices](#block-devices-configuration) documentation for fields.
 
+- `capacity_reservation_preference` (string) - Set the preference for using a capacity reservation if one exists.
```
[the remark this answers is not in the document]
Comment (mihok) [reply docs-partials/builder/ebs/Config-not-required.mdx:12]: I'll update it! Thanks

**Review by nywilken — APPROVED** [review]
@mihok thanks for making the update here. Code wise this looks great. We are in the process of addressing a regression. So I will hold on merging this today but will work on getting it out next week if there are no issues with getting out alongside the regression.

File: builder/common/step_run_source_instance.go [hunk 2]
```diff
@@ -229,6 +232,24 @@ func (s *StepRunSourceInstance) Run(ctx context.Context, state multistep.StateBa
 		}
 	}
 
+	if s.CapacityReservationPreference != "" {
```
Comment (lbajolet-hashicorp) [inline builder/common/step_run_source_instance.go:235]: If it's not empty, should we check that the value is either `none` or `open`? Or will a request fail when we set something invalid?

Comment (lbajolet-hashicorp) [reply builder/common/step_run_source_instance.go:235]: Also, I imagine it being unset is the same as setting it to `none`? If not we should probably replace the empty value with `none` and set it in the configuration

**Review by lbajolet-hashicorp — APPROVED** [review]
LGTM overall, just have a couple questions regarding `CapacityReservationPreference` regarding the documentation vs. implementation, once addressed, we can merge this!

File: builder/common/step_run_source_instance.go (hunk 2 above)
Comment (nywilken) [reply builder/common/step_run_source_instance.go:235]: Ah good catch! Thanks for the review.

**Comment by nywilken** [comment]
@mihok thanks again for all the work on this  contribution. I'll work with @lbajolet-hashicorp to get the requested changes in the PR and rebased onto the latest main. So that we can get this merged. Please let me know if you have any objections to making the added changes to your branch. Cheers!

**Comment by lbajolet-hashicorp** [comment]
Update: going through old PRs, this one was left open for a while, apologies for this @mihok.

This is now rebased on top of the current main, and I've taken the liberty to add some checks during `Prepare` to make sure the values provided to the `capacity_reservations_preference` option are valid.

With this, @nywilken when you have time, could you do a last review pass, and we can consider merging this if the code looks good to you.

@mihok: I had to rebase and fix some conflicts, so to be sure, could you do a quick pass on the final code to make sure it still is conformant to your initial proposal?

Thanks again for opening this, and sorry for taking so long to address the remaining problems.

Here's to hoping this can get merged quickly!

**Pull request merged**